odl/operator/pspace_ops.py

```diff
@@ -562,6 +628,286 @@ def adjoint(self):
+            Input vector that should be used in the reduction
+        out : range element, optional
+            output vector to write result to
+
+
+        Parameters
+        ----------
+        x : domain element
+            Input vector to be evaluated by operators
+        out : range element, optional
+            output vector to write result to
+
+        Returns
+        -------
+        out : range element
+            Sum of operators evaluated in point
+
+        Examples
+        --------
+        >>> import odl
+        >>> I = odl.IdentityOperator(odl.Rn(3))
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op([[1.0, 2.0, 3.0], [4.0, 6.0, 8.0]])
+        Rn(3).element([9.0, 14.0, 19.0])
+        """
+        if out is None:
+            return self.prod_op(x)[0]
+        else:
+            wrapped_out = self.prod_op.range.element([out], cast=False)
+            return self.prod_op(x, out=wrapped_out)
+
+    def derivative(self, x):
+        """Derivative of the reduction operator.
+
+        Parameters
+        ----------
+        x : domain element
+            The point to take the derivative in
+
+        Returns
+        -------
+        adjoint : linear `BroadcastOperator`
+            The derivative
+
+        Examples
+        --------
+        >>> import odl
+        >>> r3 = odl.Rn(3)
+        >>> I = odl.IdentityOperator(r3)
+        >>> x = r3.element([1.0, 2.0, 3.0])
+        >>> y = r3.element([4.0, 6.0, 8.0])
+
+        Example with linear operator (derivative is itself)
+
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+        >>> op.derivative([x, y])([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+
+        Example with affine operator
+
+        >>> residual_op = odl.ResidualOperator(I, r3.element([1, 1, 1]))
+        >>> op = ReductionOperator(residual_op, 2 * residual_op)
+
+        Calling operator gives offset by [3, 3, 3]
+
+        >>> op([x, y])
+        Rn(3).element([6.0, 11.0, 16.0])
+
+        Derivative of affine operator does not have this offset
+
+        >>> op.derivative([x, y])([x, y])
+        Rn(3).element([9.0, 14.0, 19.0])
+        """
+        return ReductionOperator(*[op.derivative(xi)
+                                   for op, xi in zip(self.operators, x)])
+
+    @property
+    def adjoint(self):
+        """Adjoint of the reduction operator.
+
+        Returns
+        -------
+        adjoint : linear `BroadcastOperator`
+            The adjoint
+
+        Examples
+        --------
+        >>> import odl
+        >>> I = odl.IdentityOperator(odl.Rn(3))
+        >>> op = ReductionOperator(I, 2 * I)
+        >>> op.adjoint([1, 2, 3])
+        ProductSpace(Rn(3), 2).element([
+            [1.0, 2.0, 3.0],
+            [2.0, 4.0, 6.0]
+        ])
+        """
+        return BroadcastOperator(*[op.adjoint for op in self.operators])
+
+
+def diagonal_operator(operators, dom=None, ran=None):
```

Comment:
Dont quite get it, should we add a `operators` property to `ProductSpaceOperator`? Do we know if there is a well defined order to these?